File: GARI_Umbria_2026-08-03/GiasNG/src/app/Model/anagrafiche/CentroAziendale.ts
import { TipologiaSede} from '../metaschema/TipologiaSede';
import { TitoloDiPossesso} from '../metaschema/TitoloDiPossesso';
import { OrientamentoTecnicoEconomico} from '../metaschema/OrientamentoTecnicoEconomico';
import { BioTipoAttivita } from '../metaschema/BioTipoAttivita';
import { BioOrganismoDiControllo } from '../metaschema/BioOrganismoDiControllo';
import { CatastoCentroAziendale } from './CatastoCentroAziendale';
import { CodiciAnagrafeValori } from './CodiciAnagrafeValori';
import { IndirizzoAssociato } from './addresses/IndirizzoAssociato';
import { IntervalloTemporale } from './IntervalloTemporale';
import { RubricaVoci} from './RubricaVoci';
import { Type } from '@angular/core';
import { DropdownList } from 'gias-kendo-grid';
import { formatDate } from '@angular/common';
import { CentroAziendaleEsternoCollegato } from '../metaschema/CentroAziendaleEsternoCollegato';

export type PKCentroAziendale = typeof CentroAziendale.PK.prototype;
export class CentroAziendale {
  primaryKey: PKCentroAziendale;
  nome: string;                                           // Dati Centro
  tipologia: TipologiaSede;                               // Dati Centro
  titolo_Di_Possesso: TitoloDiPossesso;                   // Dati Centro
  indirizzi: IndirizzoAssociato[];                        // Indirizzo
  lat: number;                                            // Geolocalizzazione
  lng: number;                                            // Geolocalizzazione
  rubricaVoci: RubricaVoci[];                             // Rubrica
  catastoCentroAziendale: CatastoCentroAziendale[];
  codici: CodiciAnagrafeValori[];
  validita: IntervalloTemporale;
  orientamentoTecnicoEconomico: OrientamentoTecnicoEconomico[];
  bioTipoAttivita: BioTipoAttivita;
  bioOrganismoDiControllo: BioOrganismoDiControllo;
  centroAziedaleEsternoCollegato: CentroAziendaleEsternoCollegato;
  flag_cancellazione: boolean;

  public static PK = class CentriPK {
    codice: number;
    partitaIva: string;

    constructor(codice: number, partitaIva: string) {
      this.codice = codice;
      this.partitaIva = partitaIva;
    }
  };

  constructor(primaryKey: PKCentroAziendale) {
    this.primaryKey = primaryKey;
    this.flag_cancellazione = false;
  }

  static Empty(piva: string) {
    return this.getEmptyCentroAziendale(piva);
  }

  static getEmptyCentroAziendale(piva: string) {
    const centro: CentroAziendale = {
      bioOrganismoDiControllo: {codice: '0', descrizione: '' },
      centroAziedaleEsternoCollegato: {codice: '0', descrizione: '' },
      bioTipoAttivita: {codice: ' ', descrizione: '' },
      catastoCentroAziendale: null,
      codici: [],
      flag_cancellazione: false,
      indirizzi: [],
      lat: 0,
      lng: 0,
      nome: '',
      orientamentoTecnicoEconomico: [],
      primaryKey: {
        codice: 0,
        partitaIva: piva
      },
      rubricaVoci: [],
      tipologia: {
        codice: 102,
        descrizione: null,
      },
      titolo_Di_Possesso: { codice: 0, descrizione: ''},
      validita: new IntervalloTemporale(new Date(), new Date())
    };
    let indirizzo: IndirizzoAssociato = {
      flag_cancellazione: false,
      indirizzo: {
        cap: '00000',
        codice: 0,
        flag_cancellazione: false,
        frazione: '',
        istatComune: {
          reg:'000',
          cap: '00000',
          codiceBelfiore: '',
          com: '000',
          comuni_prov: '00',
          localita: '',
          prov: '000',
          validita: null
        },
        note: '',
        stato: {
          codiceNumerico: null,
          codiceAlpha3: null,
          codice: 'IT',
          descrizione: 'Italia',
          gestioneGerarchia: 0
        },
        via: '',
        geolocation: {lat: 0, lng: 0, isValid: false}
      },
      tipo_Indirizzo: 1
    };
    centro.indirizzi.push(indirizzo);

    return centro;
  }

}

export class CentroAziendaleNG {

  public Centro: CentroAziendale;
  public DatiVisibili: CentroDropdownLists;
}

export class CentroDropdownLists {
  public CodiceOperatore: string;
  public Tipologia: AspxDropdown;
  public TitoloPossesso: AspxDropdown;
  public Stati: AspxDropdown;
  public Comune: AspxDropdown;
  public Provincie: AspxDropdown;
  public Codici: AspxDropdown;
  public TipoAttivita: AspxDropdown;
  public OrganismiControllo: AspxDropdown;
  public CentroAziendaleEsternoCollegato: AspxDropdown;
  public Otes: AspxDropdown;
}

export class CentroAziendaleDropdowns {
  public Tipologia: AspxDropdown;
  public TitoloPossesso: AspxDropdown;
  public Stati: AspxDropdown;
  public Comune: AspxDropdown;
  public Provincie: AspxDropdown;
  public Codici: AspxDropdown;
  public TipoAttivita: AspxDropdown;
  public OrganismiControllo: AspxDropdown;
  public CentroAziendaleEsternoCollegato: AspxDropdown;
  public Otes: AspxDropdown;
}

export type AspxDropdown = {
  Items: AspxDropdownItem[];
}
export type AspxDropdownItem = {
  Enabled: boolean;
  Selected: boolean;
  Text: string;
  Value: string;
}

File: GARI_Umbria_2026-08-03/DotNet/AgronicaCoreStd/AgronicaCoreModelsSTD/anagrafiche/CentroAziendale.cs
using AgronicaCoreModelsSTD.analisi;
using AgronicaCoreModelsSTD.metaschema;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Text;

namespace AgronicaCoreModelsSTD.anagrafiche
{
    public class RifCentroAziendale : RifImpresa
    {
        public int saCod { get; set; }

        // Costruttore vuoto per consentire deserializzazioni della classe
        public RifCentroAziendale() : base()
        {
            saCod = 0;
        }

        public RifCentroAziendale(string _piva, int _saCod) : base(_piva)
        {
            saCod = _saCod;
        }

        public String getCentroAziendaleFullId(string separator = "_")
        {
            return $"{partitaIva}{separator}{saCod}";
        }
    }

    public class CentroAziendale 
    {
        // public int Id { get; set; }
        //public int codice { get; set; }
        //public Impresa impresa { get; set; }
        public PK primaryKey { get; set; }
        public string nome { get; set; }

        public TipologiaSede tipologia { get; set; }
        public TitoloDiPossesso titolo_Di_Possesso { get; set; }
        public decimal lat { get; set; }
        public decimal lng { get; set; }

        public List<Appezzamento> appezzamenti { get; set; }
        public List<Fabbricato> fabbricati { get; set; }

        public List<CatastoCentroAziendale> catastoCentroAziendale { get; set; }
        public List<CodiciAnagrafeValori> codici { get; set; }

        public List<IndirizzoAssociato> indirizzi { get; set; }

        public List<RubricaVoci> rubricaVoci { get; set; }

        public IntervalloTemporale validita { get; set; }

        public List<OrientamentoTecnicoEconomico> orientamentoTecnicoEconomico { get; set; }

        public BioTipoAttivita bioTipoAttivita { get; set; }

        public BioOrganismoDiControllo bioOrganismoDiControllo { get; set; }

        public CentroAziendaleEsternoCollegato centroAziendaleEsternoCollegato { get; set; }

        public bool flag_cancellazione { get; set; }

        public CentroAziendale(PK primaryKey)
        {
            this.primaryKey = primaryKey;
            this.flag_cancellazione = false;
        }
        public class PK
        {
            public int codice { get; set; }
            public string partitaIva { get; set; }

            public PK(int codice, string partitaIva)
            {
                this.codice = codice;
                this.partitaIva = partitaIva;
            }

            public PK()
            {

            }
        }

        public CentroAziendale()
        {

        }

        public RifCentroAziendale getRiferimento()
        {
            return new RifCentroAziendale(this.primaryKey.partitaIva, this.primaryKey.codice);
        }
    }
}
